File: codingInterview/leetcode/CountGoodNodesInBinaryTree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 *
 * Given a binary tree root, a node X in the tree is named good if in the path from root to X there are no nodes with a value greater than X.
 */
void recursiveFindGoodNodes(struct TreeNode* current, int maxValue, int* count) {
    if (current == NULL)
    {
        return;
    }
    else
    {
        if (current -> val > maxValue) 
        {
            (*count)++;
            maxValue = current -> val;
        }
        else if(current -> val == maxValue)
        {
            (*count)++;
        }
        else
        {
            ;
        }
        recursiveFindGoodNodes(current -> left, maxValue, count);
        recursiveFindGoodNodes(current -> right, maxValue, count); 
    }
}

int goodNodes(struct TreeNode* root){
    int count = 0;
    recursiveFindGoodNodes(root, root -> val, &count);
    return count;
}

File: codingInterview/leetcode/CountGoodNodesInBinaryTree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int GoodNodes(TreeNode root) {
        int count = 0;
        RecursiveCountGoodNodes(root, root.val, ref count);
        
        return count;
    }
    
    public void RecursiveCountGoodNodes(TreeNode current, int maxValue, ref int count)
    {
        if (current == null)
        {
            return;
        }
        
        if (current.val > maxValue)
        {
            maxValue = current.val;
            count++;
        }
        else if (current.val == maxValue)
        {
            count++;
        }
        else
        {
            ;
        }
        RecursiveCountGoodNodes(current.left, maxValue, ref count);
        RecursiveCountGoodNodes(current.right, maxValue, ref count);
    }
}
